flux/mpibind_plugin.c

```diff
@@ -0,0 +1,455 @@
+#include <flux/shell.h>
+#include <hwloc.h>
+#include <jansson.h>
+#include <mpibind.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#define DEBUG 0
+
+/* mpibind.c - flux-shell mpibind plugin
+ *
+ *  flux-shell plugin to implement task binding and affinity using libmpibind.
+ *
+ *  OPTIONS
+ *
+ *  {
+ *    "disable":int,
+ *    "verbose":int,
+ *    "smt":int,
+ *    "greedy":int,
+ *    "gpu_optim":int,
+ *  }
+ *
+ *  e.g. to disable mpibind plugin run with `-o mpibind.disable`
+ *
+ *  OPERATION
+ *
+ *  In shell.init callback, parse any
+ *  relevant shell options, and if disabled is set, immediately return.
+ *  Otherwise, gather number of local tasks and assigned core list and call
+ *  mpibind(3) to generate mapping. Register for the 'task.exec' and
+ *  'task.init' callback and pass the mpibind mapping to this function for use.
+ */
+
+/*  The name of this plugin. To completely replace the shell's internal
+ *   affinity module, set to "affinity", otherwise choose a different name
+ *   such as "mpibind". If you don't overwrite the affinity module, you'll 
+ *   have to disable the module from within this plugin (see mpibind_init)
+ */
+#define PLUGIN_NAME "mpibind"
+
+/*  Return task id for a shell task
+ */
+static int flux_shell_task_getid (flux_shell_task_t *task)
+{
+    int id = -1;
+    if (flux_shell_task_info_unpack (task, "{ s:i }", "localid", &id) < 0)
+        return -1;
+    return id;
+}
+
+/*  Return the current task id when running in task.* context.
+ */
+static int get_taskid (flux_plugin_t *p)
+{
+    flux_shell_t *shell;
+    flux_shell_task_t *task;
+
+    if (!(shell = flux_plugin_get_shell (p)))
+        return -1;
+    if (!(task = flux_shell_current_task (shell)))
+        return -1;
+    return flux_shell_task_getid (task);
+}
+
+/*  Run hwloc_topology_restrict() with common flags for this module.
+ */
+static int topology_restrict (hwloc_topology_t topo, hwloc_cpuset_t set)
+{
+    if (hwloc_topology_restrict (topo, set, 0) < 0)
+        return (-1);
+    return (0);
+}
+
+/*  Restrict hwloc topology object to current processes binding.
+ */
+static int topology_restrict_current (hwloc_topology_t topo)
+{
+    int rc = -1;
+    hwloc_bitmap_t rset = hwloc_bitmap_alloc ();
+    if (!rset || hwloc_get_cpubind (topo, rset, HWLOC_CPUBIND_PROCESS) < 0)
+        goto out;
+    rc = topology_restrict (topo, rset);
+out:
+    if (rset)
+        hwloc_bitmap_free (rset);
+
+#if DEBUG
+    int num = hwloc_get_nbobjs_by_type(topo, HWLOC_OBJ_OSDEV_GPU);
+    fprintf(stderr, "Number of visible gpu devices: %d\n", num);
+#endif
```

Comment:
What function could we use to print to the user's console when implementing a `verbose` option, for example? Would that be `shell_log`?

Reply:
For the `task.exec` callback, you can print to stderr directly (since standard I/O redirection has been set up at this point). For all other plugin callbacks you can use `shell_log()`, which would always be displayed in job output. If you use `shell_debug()` or `shell_trace()`, then messages are only displayed if the user ran with `-o verbose` or `-o verbose=2`.